### PR TITLE
Exclude pages from having GTM snippet applied to it

### PR DESCRIPTION
**Pull request by estrada9166**
#9373

### CODE REVIEW

File: packages/gatsby-plugin-google-tagmanager/src/gatsby-ssr.js [hunk 1]
```diff
@@ -2,13 +2,24 @@ import React from "react"
...
-  { setHeadComponents, setPreBodyComponents },
+  { setHeadComponents, pathname, setPreBodyComponents },
   pluginOptions
 ) => {
   if (
     process.env.NODE_ENV === `production` ||
     pluginOptions.includeInDevelopment
   ) {
+    if (typeof pluginOptions.exclude !== `undefined`) {
+      const minimatch = require(`minimatch`)
```
Comment (pieh) [inline packages/gatsby-plugin-google-tagmanager/src/gatsby-ssr.js]: Please import at the top of file

File: packages/gatsby-plugin-google-tagmanager/src/gatsby-ssr.js [hunk 2]
```diff
@@ -2,13 +2,24 @@ import React from "react"
...
+      const excludedPaths = pluginOptions.exclude
+      // loop all the excluded paths to validate if the pattern match with the 
+      // actual pathname
+      for (let i = 0; i < excludedPaths.length; i++) {
+        const exclude = excludedPaths[i]
+        if (minimatch(pathname, exclude)) return
+      }
```
Comment (pieh) [inline packages/gatsby-plugin-google-tagmanager/src/gatsby-ssr.js]: maybe something like:
```js
if (pluginOptions.exclude.some(excludedPath => minimatch(pathname, excludedPath))) {
  // bail early if current path is excluded
  return
}
```

**Comment by estrada9166** [comment]
@pieh That's a valid point! I made the changes based on [this comment](https://github.com/gatsbyjs/gatsby/issues/9373#issuecomment-432938534) what do you think?

**Comment by pieh** [comment]
Was just reading on tag manager because I didn't used it yet and it seems like any configuration should be done on tag manager side and not in plugin here. With this approach if someone starts with excluded site and navigate to not excluded site you won't get events about it. Can you blacklist paths in google tagmanager console (or whatever it is called)?

**Comment by pieh** [comment]
> Is there any way to exclude a page from having the GTM snippet applied to it?

So this PR definitely does this. But because of answer below I'm not sure if doing this will show result that you would expect by doing this.

> The question is.. why it is going to exclude the events if `starts with excluded site and navigate to not excluded site`?

When Gatsby javascript bundle is loaded your page become Single Page App and when you click on links you don't load html from other pages (you load js resources). That's why you would never load GTM script if you would start with page that excludes GTM

**Comment by estrada9166** [comment]
@pieh Thanks for answering!! Now I get it... but my last question (I think so...) I'm using this package (with the new changes) locally and exclude `/`... so I run `npm run build`(`gatsby build`) and then `npm run serve`(`gatsby serve`)... I visit `http://localhost:9000/` and check the source code and there is no script for GTM and then I click the link to `about` and check the source code again and the is the script for GTM....

What I'm doing wrong?

Thanks!!

**Comment by pieh** [comment]
Do you use `gatsby-link` or `<a>` tags to navigate?

**Comment by estrada9166** [comment]
I use `gatsby-link`

**Comment by estrada9166** [comment]
@pieh you can check it [here](http://far-debt.surge.sh/) I just deploy it with surge

**Comment by estrada9166** [comment]
also, I just added a `console.log()` inside the `if` and another one outside it... and this was the result when I build
```
> gatsby-starter-hello-world@ build /Users/alejandroestrada/Documents/Code/Learn/gatsby/tutorial-part-four
> gatsby build

success open and validate gatsby-configs — 0.058 s
success load plugins — 0.160 s
success onPreInit — 1.090 s
success delete html and css files from previous builds— 0.030 s
success initialize cache — 0.004 s
success copy gatsby files — 0.077 s
success onPreBootstrap — 0.008 s
success source and transform nodes — 0.062 s
success building schema — 0.197 s
success createPages — 0.038 s
success createPagesStatefully — 0.014 s
success onPreExtractQueries — 0.000 s
success update schema — 0.113 s
success extract queries from components — 0.052 s
success run graphql queries — 0.016 s — 3/3 208.86 queries/second
success write out page data — 0.004 s
success write out redirect data — 0.001 s
success onPostBootstrap — 0.001 s

info bootstrap finished - 5.639 s

success Building production JavaScript and CSS bundles— 3.698 s
⠁ Building static HTML for pagesIncluded file: -> /offline-plugin-app-shell-fallback/
⠂ Building static HTML for pagesIncluded file: -> /sweet-pandas-eating-sweets/
Included file: -> /pandas-and-bananas/
Included file: -> /about/
Excluded file: -> /
{ allFile:
   { edges:
      [ [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object] ] } }
Included file: -> /my-files/
success Building static HTML for pages — 0.645 s — 6/625.60 pages/second
Generated public/sw.js, which will precache 10 files, totaling 265448 bytes.
info Done building in 10.08 sec
```

**Comment by pieh** [comment]
So I just checked your link - if you start on homepage (`/` -> excluded) and click `about` (which shouldn't be excluded), the tagmanager will not show up.

If you start on about (`/about` -> included) and click on "Pandas Eating Lots" which will navigate to homepage- tagmanager will be there and send events (if history trigger is setup in tagmanager).

So this is exactly what I mean - excluding script from rendered html is not solution here. You need to configure tag manager in google console

**Comment by estrada9166** [comment]
@pieh sorry my misunderstood.. but if I start on `/` and the visit `about` and check the source code I can see the GTM script

**Comment by pieh** [comment]
how are you checking source code? with devtools? or with "view source code"?

**Comment by pieh** [comment]
if you check with "view source code" you will see static html - but this is only used on initial visit and we don't use it when you switch pages

**Comment by estrada9166** [comment]
@pieh [here](https://www.useloom.com/share/c22ef7a8a66e4918ac012d71d597baab) is a quick video of how I'm doing it!

**Comment by pieh** [comment]
Yes - you are checking `public/index.html` and `public/about/index.html` content this way and tag manager will be in `public/about/index.html` - but this is not how gatsby work when you switch pages - you need to use devtools to check actual DOM, because when you go from `/` to `/about` using gatsby link - you don't actually load `public/about/index.html` file.

**Comment by estrada9166** [comment]
ohhh... @pieh I appreciate all your help and comments!! Thank you so much! I think we can close this PR and clarify on the issue that GTM should be handled on the google console instead of here!!

**Pull request closed without merging**